File: concepts/09-monitoring/python/logging.py
"""
Structured Logging Implementation
Demonstrates structured logging with different log levels
"""
from enum import IntEnum
from typing import Dict, Any, List, Optional
from datetime import datetime
import json
import secrets


class LogLevel(IntEnum):
    """Log levels"""
    DEBUG = 0
    INFO = 1
    WARN = 2
    ERROR = 3
    FATAL = 4


class StructuredLogger:
    """Structured logger implementation"""
    
    def __init__(self, config: Dict[str, Any] = None):
        config = config or {}
        self.service_name = config.get('service_name', 'app')
        self.environment = config.get('environment', 'development')
        self.min_level = config.get('min_level', LogLevel.INFO)
        self.logs: List[Dict[str, Any]] = []
        self.max_logs = config.get('max_logs', 1000)
    
    def _create_log_entry(self, level: LogLevel, message: str, metadata: Dict[str, Any] = None) -> Dict[str, Any]:
        """Create log entry"""
        metadata = metadata or {}
        return {
            'timestamp': datetime.now().isoformat(),
            'level': level.name,
            'levelNum': level.value,
            'service': self.service_name,
            'environment': self.environment,
            'message': message,
            **metadata,
            'traceId': metadata.get('traceId') or self._generate_trace_id(),
            'spanId': metadata.get('spanId') or self._generate_span_id()
        }
    
    def _generate_trace_id(self) -> str:
        """Generate trace ID"""
        return f'trace-{int(datetime.now().timestamp() * 1000)}-{secrets.token_hex(4)}'
    
    def _generate_span_id(self) -> str:
        """Generate span ID"""
        return f'span-{secrets.token_hex(4)}'
    
    def log(self, level: LogLevel, message: str, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Log message"""
        if level < self.min_level:
            return None
        
        metadata = metadata or {}
        log_entry = self._create_log_entry(level, message, metadata)
        
        # Store log
        self.logs.append(log_entry)
        if len(self.logs) > self.max_logs:
            self.logs.pop(0)
        
        # Output log
        self._output_log(log_entry)
        
        return log_entry
    
    def _output_log(self, log_entry: Dict[str, Any]):
        """Output log (can be overridden for different outputs)"""
        level = log_entry['level']
        timestamp = log_entry['timestamp']
        message = log_entry['message']
        
        # Color codes for console
        colors = {
            'DEBUG': '\033[36m',  # Cyan
            'INFO': '\033[32m',   # Green
            'WARN': '\033[33m',   # Yellow
            'ERROR': '\033[31m',  # Red
            'FATAL': '\033[35m'   # Magenta
        }
        reset = '\033[0m'
        
        color = colors.get(level, '')
        print(f'{color}[{timestamp}] [{level}] {message}{reset}')
        print(json.dumps(log_entry, indent=2))
    
    def debug(self, message: str, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Debug log"""
        return self.log(LogLevel.DEBUG, message, metadata)
    
    def info(self, message: str, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Info log"""
        return self.log(LogLevel.INFO, message, metadata)
    
    def warn(self, message: str, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Warn log"""
        return self.log(LogLevel.WARN, message, metadata)
    
    def error(self, message: str, error: Exception = None, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Error log"""
        error_metadata = {
            **(metadata or {}),
            'error': {
                'name': type(error).__name__,
                'message': str(error),
                'traceback': str(error.__traceback__) if error else None
            } if error else None
        }
        return self.log(LogLevel.ERROR, message, error_metadata)
    
    def fatal(self, message: str, error: Exception = None, metadata: Dict[str, Any] = None) -> Optional[Dict[str, Any]]:
        """Fatal log"""
        error_metadata = {
            **(metadata or {}),
            'error': {
                'name': type(error).__name__,
                'message': str(error),
                'traceback': str(error.__traceback__) if error else None
            } if error else None
        }
        return self.log(LogLevel.FATAL, message, error_metadata)
    
    def get_logs(self, level: Optional[LogLevel] = None, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """Get logs"""
        filtered = self.logs
        
        if level is not None:
            filtered = [log for log in filtered if log['levelNum'] >= level.value]
        
        if limit is not None:
            filtered = filtered[-limit:]
        
        return filtered
    
    def get_logs_json(self, level: Optional[LogLevel] = None, limit: Optional[int] = None) -> str:
        """Get logs as JSON"""
        return json.dumps(self.get_logs(level, limit), indent=2)
    
    def clear_logs(self):
        """Clear logs"""
        self.logs = []


def demonstrate_logging():
    """Demonstrate structured logging"""
    print('=== Structured Logging ===\n')
    
    logger = StructuredLogger({
        'service_name': 'api-service',
        'environment': 'production',
        'min_level': LogLevel.DEBUG
    })
    
    # Different log levels
    logger.debug('Debug message', {'component': 'auth', 'userId': 'user123'})
    logger.info('User logged in', {'userId': 'user123', 'ip': '192.168.1.1'})
    logger.warn('High memory usage', {'memoryUsage': '85%', 'threshold': '80%'})
    
    try:
        raise Exception('Database connection failed')
    except Exception as e:
        logger.error('Database error', e, {'database': 'users-db', 'query': 'SELECT * FROM users'})
    
    logger.fatal('System crash', Exception('Out of memory'), {'component': 'server'})
    
    # Get logs
    print('\n=== Recent Logs ===')
    recent_logs = logger.get_logs(limit=5)
    print(json.dumps(recent_logs, indent=2))
    
    # Get error logs only
    print('\n=== Error Logs ===')
    error_logs = logger.get_logs(LogLevel.ERROR)
    print(json.dumps(error_logs, indent=2))


if __name__ == '__main__':
    demonstrate_logging()

File: concepts/09-monitoring/node/logging.js
/**
 * Structured Logging Implementation
 * Demonstrates structured logging with different log levels
 */

/**
 * Log Levels
 */
const LogLevel = {
  DEBUG: 0,
  INFO: 1,
  WARN: 2,
  ERROR: 3,
  FATAL: 4
};

/**
 * Structured Logger
 */
class StructuredLogger {
  constructor(config = {}) {
    this.serviceName = config.serviceName || 'app';
    this.environment = config.environment || 'development';
    this.minLevel = config.minLevel || LogLevel.INFO;
    this.logs = [];
    this.maxLogs = config.maxLogs || 1000;
  }

  /**
   * Create log entry
   */
  createLogEntry(level, message, metadata = {}) {
    return {
      timestamp: new Date().toISOString(),
      level: Object.keys(LogLevel).find(key => LogLevel[key] === level),
      levelNum: level,
      service: this.serviceName,
      environment: this.environment,
      message,
      ...metadata,
      traceId: metadata.traceId || this.generateTraceId(),
      spanId: metadata.spanId || this.generateSpanId()
    };
  }

  /**
   * Generate trace ID
   */
  generateTraceId() {
    return `trace-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Generate span ID
   */
  generateSpanId() {
    return `span-${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Log message
   */
  log(level, message, metadata = {}) {
    if (level < this.minLevel) {
      return;
    }

    const logEntry = this.createLogEntry(level, message, metadata);
    
    // Store log
    this.logs.push(logEntry);
    if (this.logs.length > this.maxLogs) {
      this.logs.shift();
    }

    // Output log
    this.outputLog(logEntry);
    
    return logEntry;
  }

  /**
   * Output log (can be overridden for different outputs)
   */
  outputLog(logEntry) {
    const level = logEntry.level;
    const timestamp = logEntry.timestamp;
    const message = logEntry.message;
    
    // Color codes for console
    const colors = {
      DEBUG: '\x1b[36m', // Cyan
      INFO: '\x1b[32m',  // Green
      WARN: '\x1b[33m',  // Yellow
      ERROR: '\x1b[31m', // Red
      FATAL: '\x1b[35m'  // Magenta
    };
    const reset = '\x1b[0m';
    
    console.log(
      `${colors[level] || ''}[${timestamp}] [${level}] ${message}${reset}`,
      logEntry
    );
  }

  /**
   * Debug log
   */
  debug(message, metadata = {}) {
    return this.log(LogLevel.DEBUG, message, metadata);
  }

  /**
   * Info log
   */
  info(message, metadata = {}) {
    return this.log(LogLevel.INFO, message, metadata);
  }

  /**
   * Warn log
   */
  warn(message, metadata = {}) {
    return this.log(LogLevel.WARN, message, metadata);
  }

  /**
   * Error log
   */
  error(message, error = null, metadata = {}) {
    const errorMetadata = {
      ...metadata,
      error: error ? {
        name: error.name,
        message: error.message,
        stack: error.stack
      } : null
    };
    return this.log(LogLevel.ERROR, message, errorMetadata);
  }

  /**
   * Fatal log
   */
  fatal(message, error = null, metadata = {}) {
    const errorMetadata = {
      ...metadata,
      error: error ? {
        name: error.name,
        message: error.message,
        stack: error.stack
      } : null
    };
    return this.log(LogLevel.FATAL, message, errorMetadata);
  }

  /**
   * Get logs
   */
  getLogs(level = null, limit = null) {
    let filtered = this.logs;
    
    if (level !== null) {
      filtered = filtered.filter(log => log.levelNum >= level);
    }
    
    if (limit !== null) {
      filtered = filtered.slice(-limit);
    }
    
    return filtered;
  }

  /**
   * Get logs as JSON
   */
  getLogsJSON(level = null, limit = null) {
    return JSON.stringify(this.getLogs(level, limit), null, 2);
  }

  /**
   * Clear logs
   */
  clearLogs() {
    this.logs = [];
  }
}

/**
 * Request Logger Middleware
 */
class RequestLogger {
  constructor(logger) {
    this.logger = logger;
  }

  /**
   * Log HTTP request
   */
  logRequest(req, res, startTime) {
    const duration = Date.now() - startTime;
    const metadata = {
      method: req.method,
      path: req.path,
      statusCode: res.statusCode,
      duration,
      userAgent: req.headers['user-agent'],
      ip: req.ip || req.connection.remoteAddress
    };

    if (res.statusCode >= 500) {
      this.logger.error('Request failed', null, metadata);
    } else if (res.statusCode >= 400) {
      this.logger.warn('Request error', metadata);
    } else {
      this.logger.info('Request completed', metadata);
    }
  }

  /**
   * Create middleware function
   */
  middleware() {
    return (req, res, next) => {
      const startTime = Date.now();
      const traceId = this.logger.generateTraceId();
      
      // Add trace ID to request
      req.traceId = traceId;
      
      // Log request start
      this.logger.info('Request started', {
        method: req.method,
        path: req.path,
        traceId
      });

      // Log response when finished
      res.on('finish', () => {
        this.logRequest(req, res, startTime);
      });

      next();
    };
  }
}

// Example usage
function demonstrateLogging() {
  console.log('=== Structured Logging ===\n');

  const logger = new StructuredLogger({
    serviceName: 'api-service',
    environment: 'production',
    minLevel: LogLevel.DEBUG
  });

  // Different log levels
  logger.debug('Debug message', { component: 'auth', userId: 'user123' });
  logger.info('User logged in', { userId: 'user123', ip: '192.168.1.1' });
  logger.warn('High memory usage', { memoryUsage: '85%', threshold: '80%' });
  
  try {
    throw new Error('Database connection failed');
  } catch (error) {
    logger.error('Database error', error, { database: 'users-db', query: 'SELECT * FROM users' });
  }

  logger.fatal('System crash', new Error('Out of memory'), { component: 'server' });

  // Get logs
  console.log('\n=== Recent Logs ===');
  const recentLogs = logger.getLogs(null, 5);
  console.log(JSON.stringify(recentLogs, null, 2));

  // Get error logs only
  console.log('\n=== Error Logs ===');
  const errorLogs = logger.getLogs(LogLevel.ERROR);
  console.log(JSON.stringify(errorLogs, null, 2));
}

if (require.main === module) {
  demonstrateLogging();
}

module.exports = { StructuredLogger, RequestLogger, LogLevel };

